### PR TITLE
chore: Extract shared eslint config to its own package

### PR DESCRIPTION
**Pull request by scinos**
#### Changes proposed in this Pull Request

* Create package `@automattic/calypso-build-eslint` with the shared ESLint config (see https://github.com/Automattic/wp-calypso/pull/56787#issuecomment-935513016).

#### Testing instructions

* Verify all tests are green.

### CODE REVIEW

**Comment by jsnajdr** [comment]
There is an already existing `eslint-plugin-wpcalypso` package which is a swiss army knife of Calypso ESLint rules and configs. Could we move the `node.js` file to `lib/configs` there?

**Comment by scinos** [comment]
`eslint-plugin-wpcalypso` comes with its own set of bloated peer dependencies (`@babel/core` + `eslint` + some `eslint` plugins). Seems overkill to ask every consumer package to install those dependencies just to get access to a shared config file that doesn't use any of those dependencies

**Comment by jsnajdr** [comment]
Still, creating a package just for one config file, at a time when a package for "shared ESLint config files" already exists -- I think it's quite reasonable to have objections against that, isn't it? Let's talk about better solutions, if there are any.

All the peer dependencies are there because the config files reference their names as strings, e.g., `plugins: [ 'react-hooks' ]`. Does that really warrant a peer dependency? Both with and without a peer dependency declared, the consumer project will get an error if they don't install `eslint-plugin-react-hooks`. The errors will be different, that's the entire difference.

**Comment by scinos** [comment]
What are exactly the objections against a new package? 

---

It makes sense for the peer dependencies to exist in `eslint-plugin-wpcalypso` because you can't really use the package config without the peer dependencies, as all the exported eslint configs uses them. The package also exports some eslint rules, and it would be possible to use them without requiring the peer dependencies. That's a good argument to further break that package and extract the rules to its own package.

Both existing use cases (extensible eslint config and custom rules) are meant to be used by the eslint config in the root directory. In our current setup (root eslint config for all packages in the monorepo), it doesn't make sense to import `eslint-plugin-wpcalypso` in any nested eslint config file.

That's not the case with this new config. In our case, this new config is meant to be used _specially_ by nested eslint files to override the config defined in the root config file. As such, I think it is fundamentally different from `eslint-plugin-wpcalypso` and I don't think it belongs there.

Another potential approach would be to not have a package at all, and have every consumer of this shared config to require it from `../../../build-tools/eslint/node.js` or something like that. But I'd argue that's worse than a new package, as now packages are aware of the structure outside of them.

**Comment by jsnajdr** [comment]
OK, now I understand that the `node.js` config is a bit special and that the `eslint-plugin-wpcalypso` peer dependencies would be annoying in one use case and useful in another, so a new package makes sense.

About its name, `calypso-build-eslint` -- that's kind of an umbrella name suggesting there will be more ESLint stuff added into it over time. Is that the expectation? Because then it's very unclear what exactly is the difference between `calypso-build-eslint` and `eslint-plugin-wpcalypso`, because they are both umbrella packages for ESLint stuff.

> That's a good argument to further break that package and extract the rules to its own package.

So that means that #48310 was a mistake and we will be kind of reverting it, right? Because of peer dependency troubles and nothing else.

**Comment by scinos** [comment]
I agree`calypso-build-eslint` is a poor name and prone to confusion. Maybe `@automattic/calypso-eslint-overrides`? `@automattic/calypso-build-tools`? `@automattic/eslint-templates`? Do you have any suggestions? I don't think we will add lots of new configs to that package, but I'm not sure.

> So that means that #48310 was a mistake and we will be kind of reverting it, right? 

I think it depends if we think people can use our rules without using our config. It is technically possible, but is it something people actually do? Personally, I doubt it. Most eslint plugins I've seen out there combine rules and config into one package, so in that sense we are aligned with the industry. That being said, undoing #48310 is essentially harmless and transparent for most consumers if we declare the dependencies correctly.

**Comment by jsnajdr** [comment]
I like `@automattic/calypso-eslint-overrides` name the best.

> I think it depends if we think people can use our rules without using our config. It is technically possible, but is it something people actually do?

Probably not, at least no one reported any issues with having peer dependencies they don't use. It's true that including the `node.js` override into `eslint-plugin-wpcalypso` would actually cause trouble for packages that use it.

File: packages/calypso-build-eslint/index.js [hunk 1]
```diff
@@ -0,0 +1,5 @@
+const nodeConfig = require('./node');
+
+module.exports = {
+	nodeConfig,
+}
```
Comment (jsnajdr) [inline packages/calypso-build-eslint/index.js]: Maybe `exports.nodeConfig = require('./node')` is better?

File: packages/calypso-build-eslint/tsconfig.json [hunk 2]
```diff
@@ -0,0 +1,3 @@
+{
+	"extends": "@automattic/calypso-build/typescript/js-package.json"
```
Comment (jsnajdr) [inline packages/calypso-build-eslint/tsconfig.json]: Noting that this needs an update before merging because the TS stuff got extracted.

**Review by jsnajdr — APPROVED** [review]
With a different naming, like:
```js
const { nodeOverrides } = require( '@automattic/calypso-eslint-overrides' );
```
this looks good to me :+1: Preapproving because the change is basically a search&replace.

**Pull request merged**